src/System.Private.CoreLib/shared/System/Collections/Generic/Dictionary.cs

```diff
@@ -6,6 +6,7 @@
 using System.Diagnostics.CodeAnalysis;
 using System.Runtime.CompilerServices;
 using System.Runtime.Serialization;
+using Internal.Runtime.CompilerServices;
```

Comment:
We try hard to no use `Unsafe` in the common collections implementations.

Reply:
In this case, `Unsafe.As` is used to avoid expensive contravariant casts and it's the reason why this PR improves the performance of `new Dictionary<string, string>()` by 5x. Based on [a microbenchmark I performed](https://gist.github.com/svick/455c8a825915dd20a631d0dba6394064), I think the options are:

1. Keep using `Unsafe.As`. This does not have any overhead and I believe it's safe in this case.
2. Use a simple wrapper class to make the casts invariant. This is safe and fairly simple with overhead of about ~2.4 ns.
3. Keep using contravariant casts. The overhead of this option is ~70 ns, and I think that is a lot. That should improve when https://github.com/dotnet/coreclr/pull/23548 is merged, though I'm not sure by how much.

So, what should I do?

Reply:
We have been paying for this cast before this change. Is there something that makes it more expensive now?

Reply:
@jkotas A bit. This PR adds an `Unsafe` cast to `Comparer` and doing it using a regular contravariant cast would significantly decrease its performance.

Also, increasing performance of creating a `Dictionary` seems worth doing to me, since it's a very common operation and the gains are significant.

Reply:
Since we have paid for the cast before, I do not see a problem with continue doing so. It creates incentives for better optimizations. 

The security loses from using unsafe code in common collections are significant too.

Reply:
@jkotas To clarify, are you saying I should switch back to contravariant casts? What about the approach I suggested above of using a wrapper class to make the cast invariant, but still safe?

Reply:
I am not sure what you mean by a wrapper class. Can it also fix the cycle between Dictionary and globalization?

Reply:
@jkotas See [the microbenchmark I linked to above](https://gist.github.com/svick/455c8a825915dd20a631d0dba6394064), specifically the `InvariantWrapperCast` method, that doesn't use `Unsafe` and whose performance is significantly better than the regular cast.

But it doesn't matter how the cast is done, the reference to `StringComparer` will be still there. (At least I think so, I don't really understand how tree-shaking works.)